src/data/equiv/option.lean

```diff
@@ -0,0 +1,72 @@
+/-
+Copyright (c) 2021 Eric Wieser. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Eric Wieser
+-/
+import data.equiv.basic
+import control.equiv_functor
+
+/-!
+# Equivalences for `option α`
+-/
+
+namespace equiv
+
+variables {α β : Type*} (e : option α ≃ option β)
+
+private def remove_none_aux (x : α) : β :=
+if h : (e (some x)).is_some
+  then option.get h
+  else option.get $ show (e none).is_some, from
+  begin
+    rw ←option.ne_none_iff_is_some,
+    intro hn,
+    rw [option.not_is_some_iff_eq_none, ←hn] at h,
+    simpa only using e.injective h,
+  end
+
+private lemma remove_none_aux_some {x : α} (h : ∃ x', e (some x) = some x') :
+  some (remove_none_aux e x) = e (some x) :=
+by simp [remove_none_aux, option.is_some_iff_exists.mpr h]
+
+private lemma remove_none_aux_none {x : α} (h : e (some x) = none) :
+  some (remove_none_aux e x) = e none :=
+by simp [remove_none_aux, option.not_is_some_iff_eq_none.mpr h]
+
+private lemma remove_none_aux_inv (x : α) : remove_none_aux e.symm (remove_none_aux e x) = x :=
+option.some_injective _ begin
+  cases h1 : e.symm (some (remove_none_aux e x)); cases h2 : (e (some x)),
+  { rw remove_none_aux_none _ h1,
+    exact (e.eq_symm_apply.mpr h2).symm },
+  { rw remove_none_aux_some _ ⟨_, h2⟩ at h1,
+    simpa using h1, },
+  { rw remove_none_aux_none _ h2 at h1,
+    simpa using h1, },
+  { rw remove_none_aux_some _ ⟨_, h1⟩,
+    rw remove_none_aux_some _ ⟨_, h2⟩,
+    simp },
+end
+
+/-- Given an equivalence between two `option` types, eliminate `none` from that equivalence by
+mapping `e.symm none` to `e none`. -/
```

Comment:
Should there be a simp-lemma that witnesses this fact?

Reply:
One related simp lemma (`option_symm_apply_none_iff`) was added. The lemma that witnesses the docstring fact `some_remove_none_iff` is an iff that does not seem helpful to tag as simp. If that's not the case, let me know.